apps/prairielearn/src/lib/server-jobs.ts

```diff
@@ -27,10 +27,23 @@ interface ServerJobExecOptions {
   env?: NodeJS.ProcessEnv;
 }
 
-type ExecutionJob = Omit<ServerJob, 'execute' | 'executeInBackground'>;
-type ExecutionFunction = (job: ExecutionJob) => Promise<void>;
+export interface ServerJob {
+  error(msg: string): void;
+  warn(msg: string): void;
+  info(msg: string): void;
+  verbose(msg: string): void;
+  exec(file: string, args?: string[], options?: ServerJobExecOptions): Promise<void>;
+}
+
+export interface ServerJobExecutor {
+  jobSequenceId: string;
+  execute(fn: ServerJobExecutionFunction): Promise<void>;
+  executeInBackground(fn: ServerJobExecutionFunction): void;
+}
```

Comment:
Defining these two interfaces and switching the class name to `ServerJobImpl` is an attempt to make sure that we're careful with how we use these properties. Specifically, we want to avoid calling `exec()`/`error()`/etc. on the thing returned from `createServerJob()` and we want to avoid calling `execute()` or `executeInBackground()` inside an already-executing job. These separate interfaces allow us to achieve that.

I don't _love_ the `*Impl` name 🤷 but the alternative is to name the interface `IServerJob` to avoid a name conflict, which I don't love either!

Reply:
Yeah, `ServerJobImpl` is deliberately not exported, so this is strictly an implementation detail.